### PR TITLE
pkg/k8s: Remove non running containers from gadget tracer manager

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
Fixes https://github.com/kinvolk/inspektor-gadget/issues/314

### CODE REVIEW

**Review by alban — APPROVED** [review]
LGTM... but could I pick up the change into https://github.com/kinvolk/inspektor-gadget/pull/313 to avoid conflicts?

**Review by blanquicet — COMMENTED** [review]
I tested and it is working as expected!

I would just add that the logic is not that easy to follow so I would suggest either rename `createdChan` to make clear that it is used to notify about pod creation and also updates, or separate it into two different channels (But it would require to modify also the logic inside the podinformer).

File: pkg/gadgettracermanager/k8s/k8s.go
Comment (mauriciovasquezbernal) [reply pkg/gadgettracermanager/k8s/k8s.go]: Makes sense.

**Comment by mauriciovasquezbernal** [comment]
> I would just add that the logic is not that easy to follow so I would suggest either rename createdChan to make clear that it is used to notify about pod creation and also updates, or separate it into two different channels (But it would require to modify also the logic inside the podinformer).

https://github.com/kinvolk/inspektor-gadget/pull/313 is changing also this code. I'll make further changes once that's merged.

**Comment by mauriciovasquezbernal** [comment]
> LGTM... but could I pick up the change into #313 to avoid conflicts?

If we're able to have that ready for v0.3.1 that's fine. Otherwise I'll prefer to merge it to fix that problem ASAP.

**Review by blanquicet — APPROVED** [review]
LGTM!

**Comment by alban** [comment]
> If we're able to have that ready for v0.3.1 that's fine. Otherwise I'll prefer to merge it to fix that problem ASAP.

Ok, let's merge it now, and I'll rebase the other branch.

File: pkg/gadgettracermanager/gadgettracermanager.go
Comment (alban) [inline pkg/gadgettracermanager/gadgettracermanager.go:379]:
```suggestion
				// container had not been added, no need to remove it
```

File: pkg/gadgettracermanager/gadgettracermanager.go
Comment (alban) [inline pkg/gadgettracermanager/gadgettracermanager.go:390]: Isn't this variable `c` containing all containers, including the ones that are not running anymore? It might work because containers are first running before going into an error state, so `containerIDs` contains the container id, added line 399, but the code is confusing...

I suggest checking `nonrunning` just before `g.AddContainer` below so that containers created directly in the error state are not added in the first place. I think this could happen if the gadget pod is started after other pods are already in the error state. This check would be easier to implement if `GetNonRunningContainers` returns a `map[string]struct{}` instead of a `[]string`.

Comment (mauriciovasquezbernal) [reply pkg/gadgettracermanager/gadgettracermanager.go:390]: `PodToContainers()` only returns containers in running state. I agree that this code is not clean at all but I don't want to touch it until we merge https://github.com/kinvolk/inspektor-gadget/pull/313, after it we'll be able to so some more cleanup on this code. I propose to merge this as it's and rework the code after https://github.com/kinvolk/inspektor-gadget/pull/313 lands, what do you think?

Comment (alban) [reply pkg/gadgettracermanager/gadgettracermanager.go:390]: Ok for me.

**Comment by mauriciovasquezbernal** [comment]
Note: There were some test failures related to traceloop. I restarted the jobs twice and now they're gone. I'm merging it as I think those failures are not related to these changes.

**Pull request merged**
